### ISSUE TITLE
Deep Residual Learning for Image Recognition

### ISSUE THREAD

**Issue by linahourieh** [linahourieh/machine-learning-articles#2]
## TL;DR
Introduces the residual learning framework with evidence

### Article Link
https://arxiv.org/abs/1512.03385
### Author
Kaiming He

## Key Takeaways
* deep learning have led breakthroughs in Image classification
* DNN can integrate different level features
* Deeper NN caused Exploding/Vanishing Algo ==> solved by normalized Initiation and Intermediate normalization
* Another Problem raised ==> Degradation Problem = with increasing # of layers = saturated Accuracy then degrading Accuracy 
![Screenshot from 2022-08-26 13-25-36](https://user-images.githubusercontent.com/81252980/186893441-0c988659-0c70-4599-b7a7-e6c05b94bee0.png)
* The idea was to develop 
* These 2 networks should have at least the same accuracy.
![Untitled presentation](https://user-images.githubusercontent.com/81252980/186892993-1a6e1d58-eed8-43b5-b591-8f6e5c3fcb1c.png)
* In order to provide the ability to have identity mapping ==> residual block
![Screenshot from 2022-08-26 13-25-11](https://user-images.githubusercontent.com/81252980/186893442-7bc8487d-c5ac-4004-aac0-7ea915753d35.png)
*  If the added layers can be constructed as identity mappings, a deeper model should have a training error no more significant than its shallower counterpart.
* The degradation problem suggests that the solvers might have difficulty approximating identity mappings by multiple nonlinear layers. With the residual learning re-formulation, if identity mappings are optimal, the solver may simply drive the weights of the numerous nonlinear layers toward zero to approach identity mappings.

### Identity Block
<img width="729" alt="idblock3_kiank" src="https://user-images.githubusercontent.com/81252980/186895076-f00bc113-c51e-492b-861a-08ca4bdb28f3.png">

```python

def identity_block(X, f, filters, training=True, initializer=random_uniform):
    """
    Implementation of the identity block as defined in Figure 4
    
    Arguments:
    X -- input tensor of shape (m, n_H_prev, n_W_prev, n_C_prev)
    f -- integer, specifying the shape of the middle CONV's window for the main path
    filters -- python list of integers, defining the number of filters in the CONV layers of the main path
    training -- True: Behave in training mode
                False: Behave in inference mode
    initializer -- to set up the initial weights of a layer. Equals to random uniform initializer
    
    Returns:
    X -- output of the identity block, tensor of shape (m, n_H, n_W, n_C)
    """
    
    # Retrieve Filters
    F1, F2, F3 = filters
    
    # Save the input value. You'll need this later to add back to the main path. 
    X_shortcut = X
    
    # First component of main path
    X = Conv2D(filters = F1, kernel_size = 1, strides = (1,1), padding = 'valid', kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis = 3)(X, training = training) # Default axis
    X = Activation('relu')(X)
    
    ### START CODE HERE
    ## Second component of main path (≈3 lines)
    ## Set the padding = 'same'
    X = Conv2D(filters = F2, kernel_size = f, strides = (1,1), padding = 'same',kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis=3)(X, training = training)
    X = Activation('relu')(X) 

    ## Third component of main path (≈2 lines)
    ## Set the padding = 'valid'
    X = Conv2D(filters = F3, kernel_size = 1, strides = (1,1), padding = 'valid', kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis=3)(X, training = training)
    
    ## Final step: Add shortcut value to main path, and pass it through a RELU activation (≈2 lines)
    X = Add()([X_shortcut,X])
    X = Activation('relu')(X)
    ### END CODE HERE

    return X
```

### ConvBlock
<img width="647" alt="convblock_kiank" src="https://user-images.githubusercontent.com/81252980/186895286-7310c5bc-9459-4483-af4f-90439f38edd1.png">

```python

# GRADED FUNCTION: convolutional_block

def convolutional_block(X, f, filters, s = 2, training=True, initializer=glorot_uniform):
    """
    Implementation of the convolutional block as defined in Figure 4
    
    Arguments:
    X -- input tensor of shape (m, n_H_prev, n_W_prev, n_C_prev)
    f -- integer, specifying the shape of the middle CONV's window for the main path
    filters -- python list of integers, defining the number of filters in the CONV layers of the main path
    s -- Integer, specifying the stride to be used
    training -- True: Behave in training mode
                False: Behave in inference mode
    initializer -- to set up the initial weights of a layer. Equals to Glorot uniform initializer, 
                   also called Xavier uniform initializer.
    
    Returns:
    X -- output of the convolutional block, tensor of shape (m, n_H, n_W, n_C)
    """
    
    # Retrieve Filters
    F1, F2, F3 = filters
    
    # Save the input value
    X_shortcut = X


    ##### MAIN PATH #####
    
    # First component of main path glorot_uniform(seed=0)
    X = Conv2D(filters = F1, kernel_size = 1, strides = (s, s), padding='valid', kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis = 3)(X, training=training)
    X = Activation('relu')(X)

    ### START CODE HERE
    
    ## Second component of main path (≈3 lines)
    X = Conv2D(filters = F2, kernel_size = f, strides = (1, 1), padding='same', kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis = 3)(X, training=training)
    X = Activation('relu')(X)

    ## Third component of main path (≈2 lines)
    X = Conv2D(filters = F3, kernel_size = 1, strides = (1, 1), padding='valid', kernel_initializer = initializer(seed=0))(X)
    X = BatchNormalization(axis = 3)(X, training=training)
    
    ##### SHORTCUT PATH ##### (≈2 lines)
    X_shortcut = Conv2D(filters = F3, kernel_size = 1, strides = (s, s), padding='valid', kernel_initializer = initializer(seed=0))(X_shortcut)
    X_shortcut = BatchNormalization(axis = 3)(X_shortcut, training=training)
    
    ### END CODE HERE

    # Final step: Add shortcut value to main path (Use this order [X, X_shortcut]), and pass it through a RELU activation
    X = Add()([X, X_shortcut])
    X = Activation('relu')(X)
    
    return X
```
